### ISSUE TITLE
Deterministic (but undefined) layout

### ISSUE THREAD

**Issue by nikomatsakis** [rust-rfcs/unsafe-code-guidelines#35]
Can we say that layout is some deterministic function of a certain, fixed set of inputs? This would allow you to be sure that if you do not alter those inputs, your struct layout would not change, even if it meant that you can't predict precisely what it will be. For example, we might say that struct layout is a function of the struct's generic types and its substitutions, full stop -- this would imply that any two structs with the same definition are laid out the same. This might interfere with our ability to do profile-guided layout or to analyze how a struct is used and optimize based on that. (Some would call that a feature.)

Also, this presumably applies to enums as well as other types.

**Comment by nikomatsakis** [comment]
Key comments from #11:

@asajeffrey [writes](https://github.com/rust-rfcs/unsafe-code-guidelines/issues/11#issuecomment-426748955):

> Can we say something like we guarantee that the layout for a struct is only dependent on certain properties of its fields? Off the top of my head those are size, alignment and non-zero-ness, but there may be others. This would avoid having to propose an RFC with a language extension.

@RalfJung [writes](https://github.com/rust-rfcs/unsafe-code-guidelines/issues/11#issuecomment-428460437):

> So, after all this talk about layout guarantees for structs... could we do a little practical exercise and see if [rust-lang/rust#54922](https://github.com/rust-lang/rust/pull/54922) if within bounds of the guarantees, or exceeding them? This does some manual layout computation that is supposed to recompute the layout of a `repr(Rust)` struct with an unsized tail.



@Gankro [writes](https://github.com/rust-rfcs/unsafe-code-guidelines/issues/11#issuecomment-428803211):

> Quickly chiming in with a +1 for restricting `repr(rust)` to only reordering and still mandating natural/greedy c-style padding. This would make (align=1,size=0) types not affect layout, newtypes not affect layout, alignment predictable, and the size of heterogeneous types predictable. At the same time it would still allow us to do the only really relevant optimization of ordering the fields optimally. The fact that a theoretical field fuzzer wouldn't be allowed to be _as_ powerful as it could be is not especially concerning to me.

@gnzlbg [writes](https://github.com/rust-rfcs/unsafe-code-guidelines/issues/11#issuecomment-428881450) (in response to @asajeffrey):

> > Can we say something like we guarantee that the layout for a struct is only dependent on certain properties of its fields? Off the top of my head those are size, alignment and non-zero-ness,
> 
> I don't know. For example, consider the `repr(Rust)` `struct F { x: f32, y: f32, z: f32, w: f32 }`, under the assumption that `f32` has an alignment of `4` in the platform being targeted - the maximum alignment of a field of `F` is then `4`. Would it be ok for the compiler to increase the alignment of `F` to say, `16`, to facilitate SIMD operations?
> 
> We have been talking about guaranteeing the layout of homogenous tuples / structs, but I also imagine that this could happen somewhere else, e.g., `struct G { a: u64, x: f32, y: f32, z: f32, w: f32, b: u64 }` where `{xyzw}` get re-ordered at the front and the alignment increased to `16` to facilitate SIMD operations.

**Comment by nikomatsakis** [comment]
Sorry if I missed any.

**Comment by the8472** [comment]
* deterministic layout would disallow intentional randomization for fuzzing or defense purposes
* as written it would have to remain stable between compiler versions, e.g. to allow serialization or punning between libs compiled with different versions, which would prevent future changes. In essence the current implementation would become the implicit but ostensibly undefined standard
* this could have spooky action at a distance where upgrading a dependency which is a member of a struct suddenly causes the struct itself to change layout

So I would like to pose the counter-question: Why should this be guaranteed?

**Comment by ChrisJefferson** [comment]
I am +1 with @the8472 . My experience of C is that these kinds of limitations cause limitations are people get increasingly clever, as time goes by.

I'd actually be in favour of the compiler gaining randomisation, because (in my experience) even if something is technically allowed to vary, if it doesn't tend to change then people start to expect it being fixed.

Also randomisation seems (to me) a possibly cheap and effective anti-hacking technique -- if I can compile a program with a randomised order for every struct, it will be much harder for someone to make a repeatable hack, even if they find a suitable bug, as they won't know the layout of anything. Of course it isn't perfect by any means, but it's another layer of defense and I would want to see a good reason to throw it away.

If someone wants a well-defined fixed ABI, they can always use repr(C).

**Comment by rodrimati1992** [comment]
Deterministic layout seems appealing,especially if it is determined by the fields,not the generic parameters,

Even if layout is not deterministic,there should be ways to recover back some determinism through attributes.

One example attribute would be '#[phantomtype(C)]' which would guarantee that a generic parameter cannot affect the layout of the type,only allowing PhantomData/types with the same attribute to mention the generic parameter.
```
#[phantomtype(C)]
struct WithMetadata<T,C>{
    value:T,
    _metadata:PhantomWrapper<C>,
}

#[phantomtype(C)]
struct PhantomWrapper<C>(PhantomData<fn()->C>)
```
The `phantomtype` attribute would allow transmutes between WithMetadata<String,()> and WithMetadata<String,SomeMetadata> (the same type parameter `T`,different `C`) to be safe.

**Comment by the8472** [comment]
That's quite a special case that's not generalize well to the myriad of heterogenous structs with multiple non-ZST fields.

**Comment by rodrimati1992** [comment]
One example where it doesn't apply?
This is the very specific problem of wanting to guarantee stable layout for type with phantom type parameters.

**Comment by the8472** [comment]
Well, it would still be determined by some of its generic parameters, `T` in that case, obviously each concrete T can result in different layouts even though it's always the same generic fields being present.

**Comment by Gankro** [comment]
I stand by my exact request; I'm fine with field re-ordering being the exact thing allowed.

**Comment by the8472** [comment]
@Gankro couldn't that prevent some optimizations, e.g. reducing alignment (and thus size) when there are no references are taken on a private member?

**Comment by ubsan** [comment]
@the8472 that's legal under as-if

**Comment by RalfJung** [comment]
@Gankro We are discussing here whether we can say that layout is deterministic (with some list of inputs that may affect layout, and nothing else may). You are talking about constraining what layout may do, irrespective of its inputs. That's an entirely orthogonal discussion.

@ubsan "as-if"?

---

@rkruppe also writes in the PR:
> I am not sure this wording reserves the freedoms it want to reserve, and even if someone argued it does I would like it to be clarified. Since we did not get consensus to rule out profile-guided layout, the program source code is not all that informs layout. Not even if that includes build scripts, input data files, etc. that are used during the profiling run, since the program may not be deterministic w.r.t. these (e.g. it might have race conditions or depend on the system time or ...). So I don't think we can guarantee anything involving the words "deterministic function" and have to stick to something like "every time you invoke the compiler you may get a completely different layout".

Even with PGO, we should be able to guarantee that given the same PGO profile data, we compute the same layout. I do not know how PGO works, but I assume there's a "collection" phase spitting out some data into a file and then that file is used by the next compilation? So that file would be part of the input to the deterministic function.

**Comment by the8472** [comment]
@RalfJung even with the same input couldn't slight changes to the program plus optimization fuel change which structs get optimized and which don't?

**Comment by gnzlbg** [comment]
@rkruppe

>  Not even if that includes build scripts, input data files, etc. that are used during the profiling run, since the program may not be deterministic w.r.t. these (e.g. it might have race conditions or depend on the system time or ...). 

We might not want to think of the source code "written by the user" as part of the input to this deterministic function, but only think about the Rust that's left after macro expansion for this. If build scripts, proc macros, etc. have non-determinism, then that just produces a different "program". 

> I'm a big fan of randomized layout personally. I'd much more happily assert
that layouts can be randomized rather than try to offer any guarantees
about determinism besides those that we view are required for
interoperability.

It doesn't have to be either or. I sometimes, during debugging, want fully deterministic builds to reliably reproduce an issue, and other times, also during debugging, want fully randomized layouts to see if one bug is caused by some layout assumption. This screams to "implementation-defined" in the spec, but it would be nice to come up here with "definitions" for both cases that we can use in the current implementation.

**Comment by the8472** [comment]
I guess @RalfJung is aiming for reproducible builds, correct? Maybe the solution is to make struct layout *conditionally* deterministic.

Preconditions: no optimization fuel limit, no randomization, ???
Inputs: unordered sizes and alignments of struct members, layout-relevant attributes, PGO profile, compiler version, target specification, optimization flags

**Comment by the8472** [comment]
Another option for determinism is hashing the fully qualified name of the struct and using that as input for layout. That way a struct could retain a its layout between builds but two different structs with the same members would end up with different layouts, thus discouraging punning between `repr(Rust)` type without additional annotations.

**Comment by Gankro** [comment]
@RalfJung no I'm saying I am fine with it being non-deterministic *as long as* the non-determinism is constrained to field-reordering, with an otherwise simple and deterministic (same as repr(C)) alignment/padding algorithm.

**Comment by ChrisJefferson** [comment]
With regards alignment, when using SSE instructions I often use the macro ` _MM_ALIGN16`, which forces a struct to be 16-byte aligned, so it's contents can be efficiently loaded for use with SSE instructions. It would be nice if the compiler decided it wanted to use SSE on a struct it could add this by itself.

Restricting the size to be no bigger than `repr(C)` seems like a really tight restriction, as if a struct is already efficiently packed that strongly limits the options for re-ordering the members.

**Comment by the8472** [comment]
@gnzlbg we have said a lot about why the compiler might want to change layout and why a dev might want stable layout for debuggability. But very little has been said why unsafe code would need some specific guarantee.

@rodrimati1992 brought one example, punning between the same generic type with different phantom generic arguments. Without an annotation signalling that intent this would be ok. Without such an annotation it would prevent the compiler from making different optimization choices for different uses of monomorphized types.

What other uses do we have? @Gankro never stated how unsafe code would benefit from his requests for example.

**Comment by nikomatsakis** [comment]
An observation. If we did the following:

>  For example, we might say that struct layout is a function of the struct's generic types and its substitutions, full stop

but also included compiler settings in that equation, then we could still randomize layout, but we would be restricted that two structs with the same definition must still wind up with the same layout (which might be in any order etc). 

In any case, I am getting more attracted to the idea of pulling out certain special cases (e.g., single-field structs in #34) and leaving the broader question for later.

**Comment by the8472** [comment]
> but we would be restricted that two structs with the same definition must still wind up with the same layout (which might be in any order etc).

Which would preclude optimizations that take use of the structs into account.

**Comment by nikomatsakis** [comment]
@the8472 yes. I am ok with that, personally, but that is indeed one of the questions =)

**Comment by the8472** [comment]
Imo no such guarantee should be given because it would make layout-compatibility part of the API. You could write your own struct that matches the fields of a struct in a different crate (maybe by looking under the hood for private ones) and then rely on compatibility. Which is inherently brittle.

If no such guarantee is given and the compiler is allowed to be actively hostile towards such uses (e.g. via randomization) that would require the use of (currently non-existent) annotations and compile-time checking to allow such things which is a lot safer.

If some good optimization potential is found later and precluded due to guarantees then everyone would have to opt in with every struct where they would want that to apply. It's much nicer to have things apply to all safe uses of `repr(Rust)` by default and unsafe users should opt into any special treatment that they need, not the other way around.

So yeah, I'm leaning on the side of no-guarantees maximalism here. Unless concrete, important unsafe use-cases that cannot be reduced to special cases are spelled out here.

**Comment by arielb1** [comment]
Determinism between 2 structs with the same "definition", C-header-style, feels like an undesirable property to me. Both because of PGO and the likes, but also we'll have to include all the interesting properties of the definition somewhere (inhabitedness? that might rely on complicated factors).

Determinism of the "same" struct between 2 runs of the compiler also seems a bit hard to specify - we'll have to decide what it means for 2 runs of the compiler to be the same.

However, as a "quality-of-implementation" requirements:
1. The compiler should have a "reproducible" mode where "irrelevant" changes to the input don't change its output, as is required for reproducible builds.
2. The compiler should have a "stable" mode where "small" changes to the input don't make large changes to the output, for debugging.

But I don't think that the unsafe code guidelines should rely on such compiler flags - it just seems like a layering violation.

In any case, we *do* at least need *some* sort of layout determinism: maintain the same layout for datatypes that are subtypes of each-other (because otherwise subtyping won't work). We can add a similar sort of determinism between `Foo<X>` and `Foo<Y>` if `X` and `Y` are only used in "phantom" positions, which might be useful in unsafe code.

**Comment by the8472** [comment]
@arielb1 I still think the phantom case would benefit from annotations that require identical layout. That way we could still benefit from different types -> different use sites -> different optimizations may apply.

Otherwise I agree that the reproducible build and debugging concerns are orthogonal to guarantees for unsafe code, especially when we want non-reproducable (randomized) builds for shaking out bugs. E.g. the OpenJDK Hotspot compiler has some debugging flags that apply allowed code reordering more aggressively even when they're not beneficial to make it easier to find races and optimizer bugs.

**Comment by CAD97** [comment]
One subset I haven't seen mentioned yet is two instantiations of the same generic struct with layout-compatible types. E.g. if I have `Container<T>` (which does not use associated const data to `T`), then setting `T` to two layout-compatible types should produce two layout-compatible types.

This is a more specific subset of layout-compatibility for any `#[repr(Rust)]` with layout-compatible fields in the same order, as it _only_ applies to instantiations of the exact same generic definition.

Of course, user code utilizing layout-compatibility for library types in a semver-compatible way requires both the language support and the library to guarantee in documentation the exact guaranteed layout properties.

(Because, to be explicitly clear, my above proposed guarantee does not give `C<T>` compatible with `C<T'>` in general, because the container could use associated `type`s and/or `const`s to give differing layouts.)

**Comment by RalfJung** [comment]
The question is, what exactly is "layout-compatible"? For example, the niches would have to be equal. This means that `Cell<T>` and `T` are *not* "layout-compatible", and neither are `MaybeUninit<T>` and `T`.

**Comment by CAD97** [comment]
Ah, I thought we had defined "layout-compatible" already 😅 

For the minimal guarantee I suggest, `#[repr(transparent)]`-equivalent would still be enough to enable some useful patterns that aren't possible yet.

**Comment by hanna-kruppe** [comment]
`repr(transparent)` does not actually guarantee equal niches (`Cell` is transparent).

**Comment by bjorn3** [comment]
`Cell` is a wrapper around `UnsafeCell`, which uses a special attribute to hide niches.

**Comment by rodrimati1992** [comment]
For examples of types which are `#[repr(transparent)]` and *add* niches,there's the NonZero types and [NonNull](https://doc.rust-lang.org/std/ptr/struct.NonNull.html) in the standard library.
Example NonZero type: https://doc.rust-lang.org/std/num/struct.NonZeroU8.html

**Comment by CAD97** [comment]
`NonZero*` and `NonNull` use `#[rustc_layout_scalar_valid_range_start(1)]` as well, so they could be distinguished in that manner. `UnsafeCell` is "only" `#[lang = "unsafe_cell"]`.

**Comment by RalfJung** [comment]
`UnsafeCell` is also `#[repr(no_niche)]`.

**Comment by jonhoo** [comment]
I recently came across a case where I specifically wanted to cast a `SomeType<T>` into a `SomeType<ManuallyDrop<T>>`, and in trying to figure out whether that was safe I found this discussion. I _think_ that case works correctly in the compiler today, but I suspect that it's still an open discussion whether it should/will be guaranteed by the language to be okay (subject to the discussion here). I figured I'd just leave it as an additional use-case, and an indication that supporting that kind of cast would be nice :)

**Comment by scottmcm** [comment]
@jonhoo The most common form of that I see is things like expecting `Vec<i32>` and `Vec<u32>` to have the same layout, and I agree that there's no rule allowing that today.  And in the general case it can't hold -- imagine if `SomeType` used associated types (potentially even via specialization) to hold a different field type depending on the type parameter.

**Comment by digama0** [comment]
General case meaning what exactly? Ideally, we can get some promises about `A<T>` being layout compatible with `A<U>` for particular choices of `T,U` and all `A` (let's call this `T` and `U` being contextually compatible). If `T` and `U` don't define any associated types, is it possible for some unknown `A` to distinguish between them? I'm not sure exactly how specialization plays into this, but I can certainly imagine that C++ would let you do some trick using specialization to make a function `is_T<X>` which "evaluates" to true (or `u8`, say) if `X` is `T` and false (or `()`) if `X` is anything else. Is that what you mean?

**Comment by CAD97** [comment]
> If `T` and `U` don't define any associated types, is it possible for some unknown `A` to distinguish between them?

Yes, because the crate defining `A` could also define a new trait and implement it for `T` and `U`, giving them associated types.

We could _potentially_ say that for layout-compatible `T` and `U`, `A<T>` and `A<U>` are guaranteed layout-compatible if the layout of `A` does not depend on any generic information other than the layout of its generic parameters, but this is a very complicated property to actually define, and as I understand it, not even that simple to maintain in the face of PGO that's allowed to reorder fields.

**Comment by thomcc** [comment]
There are meaningful code-size reasons to want `Vec<i32>` to have the same layout as `Vec<u32>` too, although this depends on some deduplication occurring.

**Comment by RalfJung** [comment]
> I recently came across a case where I specifically wanted to cast a SomeType<T> into a SomeType<ManuallyDrop<T>>, and in trying to figure out whether that was safe I found this discussion. 

`ManuallyDrop` is `repr(transparent)`, so in terms of compatibility of `T` and `ManuallyDrop<T>` it doesn't get much better. (However I am not sure if it being `repr(transparent)` is a stable guarantee of the standard library; that'd be a t-libs questron.)

As noted above, however, it is hard to say what this means under arbitrary wrapper types.

> We could potentially say that for layout-compatible T and U, A<T> and A<U> are guaranteed layout-compatible if the layout of A does not depend on any generic information other than the layout of its generic parameters, but this is a very complicated property to actually define, and as I understand it, not even that simple to maintain in the face of PGO that's allowed to reorder fields.

What about making such a guarantee if `A`'s type parameter has no trait bound? I guess with specialization that could *still* use associated types? So we'd have to basically syntactically exclude associated types then.

But indeed this would rule out PGO-induced field reordering. There aren't many layout guarantees that are compatible with such strong forms of PGO.

**Comment by the8472** [comment]
> ManuallyDrop is repr(transparent), so in terms of compatibility of T and ManuallyDrop<T> it doesn't get much better.

`repr(transparent)` is not helpful though if you don't know its inner structure as that is non-public. I.e. you don't see what type it is transparent to (should be called translucent 😀). See https://github.com/rust-lang/unsafe-code-guidelines/issues/35#issuecomment-601224419

For a Wrapper<T> we would at least need a `repr(transparent(T))` annotation or better something like `repr(transmutable_from(T))` to make it explicit.
Plus the annotation is not shown in the documentation.

Also, `MaybeUninit` is `repr(transparent)` too, and yet it is very much not transmutable within some wrapper types.


I still think making promises about transmutability of `repr(Rust)` in the *general case* is a bad idea. We already have a few special carveouts such as single-field or homogenous structs and maybe a few more can be added; but for the general case explicit annotations should be required.

**Comment by RalfJung** [comment]
> Also, MaybeUninit is repr(transparent) too, and yet it is very much not transmutable within some wrapper types.

`repr(transparent)` on `union` is... different, indeed. That's why it is not stable yet.

> We already have a few special carveouts such as single-field or homogenous structs and maybe a few more can be added; but for the general case explicit annotations should be required.

To be clear, those caveouts are *proposals*, not guarantees, at this stage.

**Comment by Diggsey** [comment]
IMO, layout compatibility between `Generic<T>` and `Generic<U>` should be a property that the author of `Generic` chooses, and the language should only facilitate that.

For example, we could have an attribute `#[repr(parametric)]` that the author could apply to generic parameters to indicate that representation is parametric with respect to that parameter.

**Comment by elichai** [comment]
> IMO, layout compatibility between `Generic<T>` and `Generic<U>` should be a property that the author of `Generic` chooses, and the language should only facilitate that.
> 
> For example, we could have an attribute `#[repr(parametric)]` that the author could apply to generic parameters to indicate that representation is parametric with respect to that parameter.

Agreed, and about the repr, isn't that what safe transmute for? that way the author can make those kinds of guarantees

**Comment by jonhoo** [comment]
Yeah, I was thinking _specifically_ of the case of casting in/away `ManuallyDrop`, which is a type that "we" completely control. The more general case is (as has been demonstrated above) more complicated.

**Comment by RalfJung** [comment]
"We" do not control `SomeType` though, so it might be doing associated type shenanigans.

**Comment by jonhoo** [comment]
Mmm, good point. It would be really nice to have some rules around whether this is ever safe though, as it is a very handy cast to be able to do in particularly "weird" unsafe code. In my case, I use it because I know that I'm operating on aliased values, and I essentially want to "turn" any drop of those values into a forget in some code that I do not control (e.g., `HashMap::clear`). If the cast is not guaranteed to be safe (perhaps subject to some preconditions), then I don't know if there is an efficient alternative for what I'm trying to do.

**Comment by scottmcm** [comment]
I think it would generally be up to `SomeType` to provide a method for it if it's logical.

For example, once we have the safe transmute traits to use as a bound, I hope to get `vec.bikeshed()` to allow *O*(1) `Vec<i32> -> Vec<u32>` and `Vec<Foo> -> Vec<MaybeUninit<Foo>>`, etc.

**Comment by digama0** [comment]
Rather than a `.bikeshed()` method, couldn't that be an implementation of the very same `TransmuteFrom` trait? That way you can also O(1) transmute `Vec<Vec<i32>> -> Vec<Vec<u32>>` and so on.

**Comment by RalfJung** [comment]
> Mmm, good point. It would be really nice to have some rules around whether this is ever safe though, as it is a very handy cast to be able to do in particularly "weird" unsafe code. In my case, I use it because I know that I'm operating on aliased values, and I essentially want to "turn" any drop of those values into a forget in some code that I do not control (e.g., HashMap::clear).

Do you know `SomeType` or is it somehow user-defined? Rust doesn't have Generic Associated Types / Associated Type Constructors, but once it does one could write code that is fully generic in `SomeType`. In such code it is certainly *not* okay to perform these transmutes -- we'd have to have some trait bound which ensures that there are no "associated type shenanigans". However, Rust doesn't GAT/ATC, so I wonder how much you concretely know about that `SomeType`.

**Comment by digama0** [comment]
I saw the stream, and in @jonhoo 's situation `SomeType` is a known (but complicated) nesting of types, a few struct wrappers around a `HashMap` or `IndexMap`. So some external types are used (you have to know the implementation of `HashMap`) but it's not completely generic. (The question is basically whether `MyWrapper<HashMap<K, V>>` and `MyWrapper<HashMap<K, ManuallyDrop<V>>>` are layout compatible.) He contemplated having a map interface to generalize over the map type, and in that case it would look a bit more dire, given the discussion here. Most likely the trait would have to directly support a `ManuallyDrop`-like operation, because you can't quantify over type constructors anyway.

**Comment by scottmcm** [comment]
> couldn't that be an implementation of the very same `TransmuteFrom` trait?

That depends whether the standard library wants to make that promise.  I don't know if it's willing to commit to never tweaking how it's stored.

Perhaps it will, and the only things that will need the method are things like *O*(1) `Vec<[f32; 4]>` -> `Vec<f32>` that cannot possibly be a transmute (without changing `len` to do a division or something).

**Comment by RustyYato** [comment]
@scottmcm my [`vec-utils`](https://crates.io/crates/vec-utils) crate can do that [`VecExt::drop_and_reuse`](https://docs.rs/vec-utils/0.3.0/vec_utils/trait.VecExt.html#tymethod.drop_and_reuse).

**Comment by jonhoo** [comment]
Another example of a potentially useful transmute is between `SomeWrapper<T>` and `SomeWrapper<UnsafeCell<T>>` with the same intended use-case as the `ManuallyDrop` transmute (see also https://github.com/rust-lang/unsafe-code-guidelines/issues/258#issuecomment-734995613).

**Comment by the8472** [comment]
This was already discussed in previous comments. That transmute is incorrect in general (for arbitrary T) due to `UnsafeCell` being `#[repr(no_niche)]`, i.e. `size_of::<Wrapper<T>>() != Wrapper<UnsafeCell<T>>` for some combinations of T and Wrapper.

**Comment by jonhoo** [comment]
Continuing down this path, what about a transmute from
```rust
SomeWrapper<MyType<T, MyPrivateTypeA>>
```
to
```rust
SomeWrapper<MyType<T, MyPrivateTypeB>>
```
where `MyPrivateTypeA` and `MyPrivateTypeB` are both unit structs that implement no traits, and `MyType` holds only the `T` (and the private type in `PhantomData`) and is `repr(transparent)`?

I could be mistaken, but I think there is now no way to have an impl outside the current crate that distinguishes between the two types, and therefore the transmute _should_ be safe?

**Comment by the8472** [comment]
While `MyType`'s layout would be covered by the rule proposed in #34 ([single field structs](https://rust-lang.github.io/unsafe-code-guidelines/layout/structs-and-tuples.html#single-field-structs)) it doesn't tell us what properties `SomeWrapper` would have. So in the general case for arbitrary wrapper types you'd have to assume the compiler could make different decisions for each different `U` in  `SomeWrapper<U>`.

**Comment by jonhoo** [comment]
Hmm, that's unfortunate indeed. And it doesn't sound like you think there's any chance of that being the case subject to the discussion on this issue either without placing restrictions on the wrapper type? Essentially, I'm trying to find a way to use the type system to change the drop behavior of a type that is nested inside another data structure, but without being able to safely transmute the type "inside" the data structure, I cannot find a way to do so.  Which leads me down the path of _horrible_ hacks like https://github.com/jonhoo/rust-evmap/blob/4e83da000cfee94810eb790f25c1dc159def1b12/evmap/src/aliasing.rs#L79-L94. I was hoping there might be a way to construct the inner types in such a way that the layout of the wrapper type is guaranteed to stay the same (making the transmute safe), but it sounds like that's not the case.. hmm hmm hmm what to do.

**Comment by the8472** [comment]
> Essentially, I'm trying to find a way to use the type system to change the drop behavior of a type that is nested inside another data structure, but without being able to safely transmute the type "inside" the data structure, 

Can you reach into the data structure to touch the values? If so you could try exploiting the niche optimization instead of using the type system. It would come at a runtime cost but at least you wouldn't be suffering size overhead to carry that state as long as types have niches.

E.g. 

```rust
enum DropToggle<T> {
  On(T),
  Off(ManuallyDrop<T>)
}
```

**Comment by jonhoo** [comment]
In this case the wrapper type is a collection type (worse yet, a collection of collections), so while I _could_ walk every value to toggle the value, that operation is quite expensive as it is proportional to the size of the collection, even if only one item actually ends up being dropped. Think of something like calling `HashMap::retain` with a user-provided closure — you would need to walk all the values in the map, flip all of them, run the retain, then walk all of them again to flip them off again.

**Comment by gThorondorsen** [comment]
Prior art related to the current discussion: Haskell's [type roles](https://gitlab.haskell.org/ghc/ghc/-/wikis/roles) and the [Data.Coerce](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Coerce.html) library ([referenced in the safe transmute RFC](https://github.com/rust-lang/rfcs/blob/ee1204c23320c33437107ff222e09afd9d5fef1d/text/0000-safer-transmute.md#prior-art-haskell)).

**Comment by RalfJung** [comment]
> I was hoping there might be a way to construct the inner types in such a way that the layout of the wrapper type is guaranteed to stay the same

So, you are asking for some amount of "parametricity" on the type level. That sounds like a very reasonable request. To make this work, we need to *at least* ensure that the two types also behave identically for every trait resolution query.

Your approach of using private types that implement no traits seems to be a nice step in that direction. Unfortunately I don't have enough intuition for specialization to really say if that is enough.

**Comment by the8472** [comment]
> Prior art related to the current discussion: Haskell's type roles and the Data.Coerce library (referenced in the safe transmute RFC).

So, my haskell is weak, but if I understand type roles document correctly then it is primarily about representing the difference between exact type equality vs. layout equality based on generic parameters in the type system and then goes on to suggest that exact (norminal) type equality should be the default for all types and layout equivalence should require explicit annotations. In other words they suggest that by default there should be no promises about layout equality.


> The main question here is if we want to make this a guarantee that will be maintained in the future. 

Technically the initial comment only asks whether it is a deterministic function of a fixed set of inputs. Those inputs could contain the exact type (nominal equality in the type roles document) or a random input.
It would also have to include something to preclude ossification of the ostensibly unspecified layout (e.g. the compiler hash), otherwise people would just dump memory representation to disk and expect it to be loadable in a later build.


Questions that I think need to be answered

* Is there any value in the context of the UCG to guarantee deterministic layout where one of the inputs of the deterministic function is external lookup (PGO, layout randomization seed) and/or the full type name? It basically sounds like a non-guarantee to me.
* Do we want to keep the possibility of PGO-based layout or punning-hostile randomization open?
* Do we want to enable transmutation of inner types in _arbitrary_ containers? If so should this require annotations on the container and/or the inner type? Which other conditions would apply? In other words, under which conditions whould the function be guaranteed to be surjective?
* How do we prevent ossification if we make such guarantees?

**Comment by RalfJung** [comment]
Does `-Zrandomize-layout` satisfy *any* of the proposals made in this thread regarding determinism / parametricity of layout? AFAIK not, which probably means we should just close this issue as there's not really anything we'd like to guarantee here?

**Comment by scottmcm** [comment]
Agreed, if this is a thing it should be an opt-in `repr` because otherwise we're closing off unknown unknowns.

If we'd done this in 2017 we would have said "well same size and align of a field means same layout", and that would have been a mistake because we now use more things than just that in the layout calculations.

**Comment by RalfJung** [comment]
Okay, let's close this for now. If this discussion is to be reopened it should be with a somewhat more concrete proposal.

**Comment by the8472** [comment]
> Does -Zrandomize-layout satisfy any of the proposals made in this thread regarding determinism / parametricity of layout? 

See [tests/ui/layout/randomize.rs](https://github.com/rust-lang/rust/blob/7dc2c162b9c197aaa76a6f9e7534569537830a01/tests/ui/layout/randomize.rs) for its current behavior.

Currently it ignores any generic, maybe-unsized tail fields, which includes single-field wrappers, even if a particular instance of T might not be unsizable.

Additionally it does not inject padding at offset 0, but that might get added at some point https://github.com/rust-lang/rust/pull/97861.

Combining these properties means that currently `SingleField(T) == SingleField(U)` have the same layout even without `repr(transparent)`¹ if T and U do.
That does extend to  `Foo<SingleField<T>>` and `Foo<SingleField<U>>`, because we stop propagating the inner type information (if ?Sized).
It does not extend to  `Foo<SingleFieldA<T>>` and `Foo<SingleFieldB<T>>`, because `Foo` can distinguish the wrappers themselves.

These are current implementation details that could get tightened in the future.

Overall the rules are fairly complicated and subject to change that relying on `repr(Rust)` transmutability is on very thin ice.


¹ and as the test explains: `repr(transparent)` on a non-generic wrapper does nothing, which means `Foo<Transparent(Concrete)>` and `Foo<Concrete>` can get different layouts, due to potential associated-type-specialization of `Foo` in the future.
